nipype/pipeline/engine/nodes.py

```diff
@@ -277,23 +271,28 @@ def hash_exists(self, updatehash=False):
         # of the dictionary itself.
         hashed_inputs, hashvalue = self._get_hashval()
         outdir = self.output_dir()
-        if op.exists(outdir):
-            logger.debug('Output dir: %s', to_str(os.listdir(outdir)))
-        hashfiles = glob(op.join(outdir, '_0x*.json'))
-        logger.debug('Found hashfiles: %s', to_str(hashfiles))
-        if len(hashfiles) > 1:
-            logger.info(hashfiles)
-            logger.info('Removing multiple hashfiles and forcing node to rerun')
-            for hashfile in hashfiles:
-                os.unlink(hashfile)
         hashfile = op.join(outdir, '_0x%s.json' % hashvalue)
-        logger.debug('Final hashfile: %s', hashfile)
-        if updatehash and op.exists(outdir):
-            logger.debug("Updating hash: %s", hashvalue)
-            for file in glob(op.join(outdir, '_0x*.json')):
-                os.remove(file)
-            self._save_hashfile(hashfile, hashed_inputs)
-        return op.exists(hashfile), hashvalue, hashfile, hashed_inputs
+        logger.debug('Node hash value: %s', hashvalue)
+
+        if op.exists(outdir):
+            # Find previous hashfiles
+            hashfiles = glob(op.join(outdir, '_0x*.json'))
+            if len(hashfiles) > 1:  # Remove hashfiles if more than one found
```

Comment:
`or (hashfiles and hashfiles[0] != hashfile)`

this would ensure that two hashfiles are never left in the directory.

also we should check for any unfinished hashfiles in the directory and raise an exception. i.e. tell the user that you can update a hashfile in an unfinished hashfile situation.

Reply:
Regarding the second idea: are you sure we want to error? We don't want to just dismiss any viable hashfile and clear up the folder (if an unfinished hashfile was found)?

I'm pushing a commit for this, with error. Let me know if we just want to clear up the folder and return no hashfiles.

Reply:
BTW @satra, what do you think about passing the responsibilities of hashing/caching on to the interface?

The `run()` of the interface would have a `force_run` argument (so it can be done dynamically) and a `updatehash` to also support that option.

It'd be great to have the interfaces cache themselves.